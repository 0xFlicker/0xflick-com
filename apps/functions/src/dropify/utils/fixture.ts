import { IMetadata } from "@0xflick/models";

export function fakeMetadata(): IMetadata {
  return {
    name: "Fame Lady #0",
    description:
      "Fame Lady Society is the wrapped token for the first ever generative all-female avatar collection on the Ethereum blockchain. Yes, we are THE community who took over a project TWICE to write our own story. This is NFT history. This is HERstory. FLS are 8888 distinctive Ladies made up of millions of fierce trait combinations. Community = Everything. Commercial IP rights of each Lady NFT belong to its owner.",
    image: "0.png",
    attributes: [
      {
        trait_type: "Skin",
        value: "Black",
      },
      {
        trait_type: "Background",
        value: "Purple",
      },
      {
        trait_type: "Face Expression",
        value: "Pokerface",
      },
      {
        trait_type: "Eyebrows",
        value: "Stylish",
      },
      {
        trait_type: "Eyes",
        value: "Blue",
      },
      {
        trait_type: "Body",
        value: "Rose",
      },
      {
        trait_type: "Clothes",
        value: "Orange Sexy Dress",
      },
      {
        trait_type: "Glasses",
        value: "Lv",
      },
      {
        trait_type: "Hair",
        value: "Pink Italy",
      },
      {
        trait_type: "Ear Ring",
        value: "Green",
      },
      {
        trait_type: "Head",
        value: "Nimbus",
      },
      {
        trait_type: "Wrapped",
        value: "true",
      },
      {
        trait_type: "OG Rank",
        value: 4042,
      },
    ],
  };
}
