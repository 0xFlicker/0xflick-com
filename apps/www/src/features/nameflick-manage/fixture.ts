import { INameflickToken } from "@0xflick/models";
import { utils, Wallet } from "ethers";
import { tokenToStatusDescription } from "./utils";

const randomPoolOfAddresses = new Array(10)
  .fill(0, 0, 10)
  .map(() => Wallet.createRandom().address);
function randomAddress() {
  return randomPoolOfAddresses[
    Math.floor(Math.random() * randomPoolOfAddresses.length)
  ];
}

export const token1: INameflickToken = {
  tokenId: 1,
  metadata: {
    image: "https://image.0xflick.com/nameflick-image/example.nameflick.eth/1",
    name: "Nameflick #1",
    status: "FREE_USE",
    wrappedEns: "*.nameflick.eth",
    description:
      "Free to use. Allows you to claim and use one sub-domain on nameflick.eth",
    records: {
      abc: {
        normalized: "abc.nameflick.eth",
        ttl: 0,
        ensHash: utils.namehash("abc.nameflick.eth"),
        addresses: {
          eth: randomAddress(),
        },
        textRecord: {},
      },
    },
  },
};

export const token2: INameflickToken = {
  tokenId: 2,
  metadata: {
    image: "https://image.0xflick.com/nameflick-image/name.eth/2",
    name: "Nameflick #2",
    status: "PERSONAL_USE",
    wrappedEns: "name.eth",
    description:
      "Wrapped ENS. Allows you to set unlimited sub-domains and records on one ENS name",
    records: {
      alpha: {
        normalized: "alpha.name.eth",
        ttl: 0,
        ensHash: utils.namehash("alpha.name.eth"),
        addresses: {
          eth: randomAddress(),
        },
        textRecord: {},
      },
    },
  },
};

function* range(start: number, end: number, pad: number = 0) {
  for (let i = start; i < end; i++) {
    yield i.toString().padStart(pad, "0");
  }
}
export const token3: INameflickToken = {
  tokenId: 3,
  metadata: {
    image: "https://image.0xflick.com/nameflick-image/nft.eth/3",
    name: "Nameflick #3",
    status: "COMMUNITY_USE",
    wrappedEns: "nft.eth",
    description: tokenToStatusDescription("COMMUNITY_USE"),
    records: Object.fromEntries(
      [...range(1, 20, 3)].map((i) => [
        i,
        {
          normalized: `${i}.nft.eth`,
          ttl: 0,
          ensHash: utils.namehash(`${i}.nft.eth`),
          addresses: {
            eth: randomAddress(),
          },
          textRecord: {},
        },
      ])
    ),
  },
};

export const tokens = [token1, token2, token3];
