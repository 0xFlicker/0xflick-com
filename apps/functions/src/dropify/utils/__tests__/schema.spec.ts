import {
  List,
  makeData,
  Struct,
  StructRow,
  StructRowProxy,
  Utf8,
  vectorFromArray,
  Float64,
  Field,
  Table,
  tableFromJSON,
} from "apache-arrow";
import type { WriteStream } from "fs";
import {
  ParquetWriter as NewParquetWrite,
  ParquetSchema as NewParquetSchema,
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
} from "parquet-lite";
import {
  ParquetWriter as OldParquetWriter,
  ParquetSchema as OldParquetSchema,
} from "parquetjs";
import { Readable, Writable, Stream } from "stream";
import {
  attributeNumericStruct,
  attributeStringStruct,
  MetadataSchema,
  metadataSchema,
  propertiesStruct,
} from "../schema";

import { fakeMetadata } from "../fixture";
import {
  IMetadata,
  IAttributeString,
  IAttributeNumeric,
  IMetadataAttribute,
} from "@0xflick/models";

const ParquetWriter = NewParquetWrite as typeof OldParquetWriter;
const ParquetSchema = NewParquetSchema as typeof OldParquetSchema;

function makeStructRow(index: number) {
  const struct = makeData({
    type: new Struct<MetadataSchema>(metadataSchema.fields),
  });
  return new StructRow(struct, index) as StructRowProxy<MetadataSchema>;
}

function applyData(row: StructRowProxy<MetadataSchema>, data: IMetadata) {
  const [stringAttributes, numericAttributes] = data.attributes.reduce(
    ([stringAttributes, numericAttributes], attribute) => {
      if (typeof attribute.value === "string") {
        stringAttributes.push({
          ...attribute,
          colors: "colors" in attribute ? attribute.colors : null,
        } as IAttributeString);
      } else {
        numericAttributes.push({
          ...attribute,
          display_type:
            "display_type" in attribute ? attribute.display_type : null,
        } as IAttributeNumeric);
      }
      return [stringAttributes, numericAttributes];
    },
    [[], []] as [IAttributeString[], IAttributeNumeric[]]
  );
  row.image = data.image;
  row.description = data.description;
  row.tokenId = data.tokenId;
  row.external_url = data.external_url;
  row.name = data.name;
  row.attributes = vectorFromArray<
    Struct<{
      value: Utf8;
      trait_type: Utf8;
      colors: List<Utf8>;
    }>
  >(stringAttributes, attributeStringStruct);
  row.numericAttributes = vectorFromArray<
    Struct<{
      value: Float64;
      trait_type: Utf8;
      display_type: Utf8;
    }>
  >(numericAttributes, attributeNumericStruct);
  if (data.properties) {
    row.properties = vectorFromArray(
      Object.entries(data.properties).map(([key, value]) => ({ key, value })),
      propertiesStruct
    );
  }

  row.id = data.id?.toString() ?? null;
}

interface INormalizedAttribute {
  string_value: string | null;
  numeric_value: number | null;
  trait_type: string;
  display_type: string | null;
  colors: string[] | null;
  tokenId: string;
  contractAddress: string;
}

const normalizedAttributeSchema = new ParquetSchema({
  string_value: { type: "UTF8" },
  numeric_value: { type: "DOUBLE" },
  trait_type: { type: "UTF8" },
  display_type: { type: "UTF8" },
  colors: {
    repeated: true,
    type: "UTF8",
  },
  tokenId: { type: "UTF8" },
  contractAddress: { type: "UTF8" },
});

describe("Schema", () => {
  it("can make one", () => {
    const now = Date.now();
    console.log("Start makeData");
    const row = makeStructRow(0);
    try {
      applyData(row, fakeMetadata());
    } catch (e) {
      console.error(e);
      throw e;
    } finally {
      console.log("End makeData", Date.now() - now);
    }
    expect(row.toJSON()).toMatchSnapshot();
  });

  function normalizeAttributes(
    tokenId: string,
    contractAddress: string,
    attributes: IMetadataAttribute[]
  ): INormalizedAttribute[] {
    return attributes.map((attribute) => {
      if (typeof attribute.value === "string") {
        return {
          string_value: attribute.value,
          numeric_value: null,
          trait_type: attribute.trait_type,
          display_type: null,
          colors: "colors" in attribute ? attribute.colors : null,
          tokenId,
          contractAddress,
        };
      }
      return {
        string_value: null,
        numeric_value: attribute.value,
        trait_type: attribute.trait_type,
        display_type:
          "display_type" in attribute ? attribute.display_type : null,
        colors: null,
        tokenId,
        contractAddress,
      };
    });
  }

  it("vector from schema", () => {
    const testMetadataSchema = new Struct([
      new Field("string_value", new Utf8(), true),
      new Field("numeric_value", new Float64(), true),
      new Field("trait_type", new Utf8(), false),
      new Field("display_type", new Utf8(), true),
      new Field("colors", new List(new Field("color", new Utf8())), true),
      new Field("tokenId", new Utf8(), false),
      new Field("contractAddress", new Utf8(), false),
    ]);
    const vector = vectorFromArray(
      normalizeAttributes("1", "0x1", fakeMetadata().attributes),
      testMetadataSchema
    );
    expect(vector.toJSON()).toMatchSnapshot();
  });

  it("table from schema", async () => {
    const table = tableFromJSON(
      normalizeAttributes(
        "1",
        "0x1",
        fakeMetadata().attributes
      ) as unknown as Record<string, unknown>[]
    );
    const writableStream = new WritableStream();

    const writer = await ParquetWriter.openStream(
      normalizedAttributeSchema,
      writableStream as unknown as WriteStream
    );
    const parquetData: Buffer[] = [];
    writableStream.on("data", (data) => {
      parquetData.push(data);
    });

    return new Promise<void>(async (resolve, reject) => {
      writableStream.on("finish", () => {
        resolve();
      });
      writableStream.on("error", (error) => {
        reject(error);
      });
      for (const row of table) {
        await writer.appendRow(row);
      }
      writer.close();
    });
  });
});
