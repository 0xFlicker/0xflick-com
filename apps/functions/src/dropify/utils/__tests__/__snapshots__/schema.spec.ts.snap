// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Schema can make one 1`] = `
Object {
  "attributes": null,
  "description": null,
  "edition": null,
  "external_url": null,
  "id": null,
  "image": null,
  "name": null,
  "numericAttributes": null,
  "properties": null,
  "tokenId": null,
}
`;

exports[`Schema no schema 1`] = `
Array [
  Object {
    "colors": null,
    "contractAddress": "0x1",
    "display_type": null,
    "numeric_value": null,
    "string_value": "Black",
    "tokenId": "1",
    "trait_type": "Skin",
  },
  Object {
    "colors": null,
    "contractAddress": "0x1",
    "display_type": null,
    "numeric_value": null,
    "string_value": "Purple",
    "tokenId": "1",
    "trait_type": "Background",
  },
  Object {
    "colors": null,
    "contractAddress": "0x1",
    "display_type": null,
    "numeric_value": null,
    "string_value": "Pokerface",
    "tokenId": "1",
    "trait_type": "Face Expression",
  },
  Object {
    "colors": null,
    "contractAddress": "0x1",
    "display_type": null,
    "numeric_value": null,
    "string_value": "Stylish",
    "tokenId": "1",
    "trait_type": "Eyebrows",
  },
  Object {
    "colors": null,
    "contractAddress": "0x1",
    "display_type": null,
    "numeric_value": null,
    "string_value": "Blue",
    "tokenId": "1",
    "trait_type": "Eyes",
  },
  Object {
    "colors": null,
    "contractAddress": "0x1",
    "display_type": null,
    "numeric_value": null,
    "string_value": "Rose",
    "tokenId": "1",
    "trait_type": "Body",
  },
  Object {
    "colors": null,
    "contractAddress": "0x1",
    "display_type": null,
    "numeric_value": null,
    "string_value": "Orange Sexy Dress",
    "tokenId": "1",
    "trait_type": "Clothes",
  },
  Object {
    "colors": null,
    "contractAddress": "0x1",
    "display_type": null,
    "numeric_value": null,
    "string_value": "Lv",
    "tokenId": "1",
    "trait_type": "Glasses",
  },
  Object {
    "colors": null,
    "contractAddress": "0x1",
    "display_type": null,
    "numeric_value": null,
    "string_value": "Pink Italy",
    "tokenId": "1",
    "trait_type": "Hair",
  },
  Object {
    "colors": null,
    "contractAddress": "0x1",
    "display_type": null,
    "numeric_value": null,
    "string_value": "Green",
    "tokenId": "1",
    "trait_type": "Ear Ring",
  },
  Object {
    "colors": null,
    "contractAddress": "0x1",
    "display_type": null,
    "numeric_value": null,
    "string_value": "Nimbus",
    "tokenId": "1",
    "trait_type": "Head",
  },
  Object {
    "colors": null,
    "contractAddress": "0x1",
    "display_type": null,
    "numeric_value": null,
    "string_value": "true",
    "tokenId": "1",
    "trait_type": "Wrapped",
  },
  Object {
    "colors": null,
    "contractAddress": "0x1",
    "display_type": null,
    "numeric_value": 4042,
    "string_value": null,
    "tokenId": "1",
    "trait_type": "OG Rank",
  },
]
`;

exports[`Schema vector from schema 1`] = `
Array [
  Object {
    "colors": null,
    "contractAddress": "0x1",
    "display_type": null,
    "numeric_value": null,
    "string_value": "Black",
    "tokenId": "1",
    "trait_type": "Skin",
  },
  Object {
    "colors": null,
    "contractAddress": "0x1",
    "display_type": null,
    "numeric_value": null,
    "string_value": "Purple",
    "tokenId": "1",
    "trait_type": "Background",
  },
  Object {
    "colors": null,
    "contractAddress": "0x1",
    "display_type": null,
    "numeric_value": null,
    "string_value": "Pokerface",
    "tokenId": "1",
    "trait_type": "Face Expression",
  },
  Object {
    "colors": null,
    "contractAddress": "0x1",
    "display_type": null,
    "numeric_value": null,
    "string_value": "Stylish",
    "tokenId": "1",
    "trait_type": "Eyebrows",
  },
  Object {
    "colors": null,
    "contractAddress": "0x1",
    "display_type": null,
    "numeric_value": null,
    "string_value": "Blue",
    "tokenId": "1",
    "trait_type": "Eyes",
  },
  Object {
    "colors": null,
    "contractAddress": "0x1",
    "display_type": null,
    "numeric_value": null,
    "string_value": "Rose",
    "tokenId": "1",
    "trait_type": "Body",
  },
  Object {
    "colors": null,
    "contractAddress": "0x1",
    "display_type": null,
    "numeric_value": null,
    "string_value": "Orange Sexy Dress",
    "tokenId": "1",
    "trait_type": "Clothes",
  },
  Object {
    "colors": null,
    "contractAddress": "0x1",
    "display_type": null,
    "numeric_value": null,
    "string_value": "Lv",
    "tokenId": "1",
    "trait_type": "Glasses",
  },
  Object {
    "colors": null,
    "contractAddress": "0x1",
    "display_type": null,
    "numeric_value": null,
    "string_value": "Pink Italy",
    "tokenId": "1",
    "trait_type": "Hair",
  },
  Object {
    "colors": null,
    "contractAddress": "0x1",
    "display_type": null,
    "numeric_value": null,
    "string_value": "Green",
    "tokenId": "1",
    "trait_type": "Ear Ring",
  },
  Object {
    "colors": null,
    "contractAddress": "0x1",
    "display_type": null,
    "numeric_value": null,
    "string_value": "Nimbus",
    "tokenId": "1",
    "trait_type": "Head",
  },
  Object {
    "colors": null,
    "contractAddress": "0x1",
    "display_type": null,
    "numeric_value": null,
    "string_value": "true",
    "tokenId": "1",
    "trait_type": "Wrapped",
  },
  Object {
    "colors": null,
    "contractAddress": "0x1",
    "display_type": null,
    "numeric_value": 4042,
    "string_value": null,
    "tokenId": "1",
    "trait_type": "OG Rank",
  },
]
`;
